In [187]:
# Importing modules.
import pandas as pd

## Customer's Table

In [188]:
# Loading the 'Customers' csv into Dataframe.
customers_df = pd.read_csv('Customers.csv', index_col='CustomerKey', encoding = 'unicode_escape')
customers_df.tail(5)

Gender              Name                 City State Code  \
CustomerKey                                                             
2099600      Female    Denisa Dušková              Houston         TX   
2099618        Male  Justin Solórzano               Mclean         VA   
2099758        Male   Svend Petrussen           Wilmington         NC   
2099862      Female      Lorenza Rush            Riverside         CA   
2099937        Male  Zygmunt Kaminski  Bloomfield Township         MI   

                      State Zip Code        Country      Continent    Birthday  
CustomerKey                                                                     
2099600               Texas    77017  United States  North America   3/25/1936  
2099618            Virginia    22101  United States  North America   2/16/1992  
2099758      North Carolina    28405  United States  North America   11/9/1937  
2099862          California    92501  United States  North America  10/12/1937  
2099937            Michigan    48302  United States  North America   8/18/1965

In [189]:
# Reviewing Dataframe.
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15266 entries, 301 to 2099937
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Gender      15266 non-null  object
 1   Name        15266 non-null  object
 2   City        15266 non-null  object
 3   State Code  15256 non-null  object
 4   State       15266 non-null  object
 5   Zip Code    15266 non-null  object
 6   Country     15266 non-null  object
 7   Continent   15266 non-null  object
 8   Birthday    15266 non-null  object
dtypes: object(9)
memory usage: 1.2+ MB


- 10 columns:CustomerKey, Gender, Name, City, State Code, State, Zip Code, Country, Continenet and Birthday.

- 15266 entries.
- No null entries
- All columns are listed as object, upon observing data that does not seem to be correct.

In [190]:
# Checking for duplicate values.
customers_df.duplicated().sum()

0

- No duplicated entries.

In [191]:
# Checking unique values for Category candidates. 
customers_df.nunique()

Gender            2
Name          15118
City           8258
State Code      467
State           512
Zip Code       9505
Country           8
Continent         3
Birthday      11270
dtype: int64

- Gender, Country and Continent potential category columns.

In [192]:
# Checking memory usage.
customers_df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15266 entries, 301 to 2099937
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Gender      15266 non-null  object
 1   Name        15266 non-null  object
 2   City        15266 non-null  object
 3   State Code  15256 non-null  object
 4   State       15266 non-null  object
 5   Zip Code    15266 non-null  object
 6   Country     15266 non-null  object
 7   Continent   15266 non-null  object
 8   Birthday    15266 non-null  object
dtypes: object(9)
memory usage: 8.8 MB


- 8.8 MB of memory.

In [193]:
# Categorizing Gender Column.
customers_df['Gender'].unique()

array(['Female', 'Male'], dtype=object)

- Only two gender options.

In [194]:
customers_df['Gender'] = pd.Categorical(
    values = customers_df['Gender'], 
    categories = ['Female', 'Male'],
    ordered = True
)

In [195]:
# Categorizing Country Column.
customers_df['Country'].unique()

array(['Australia', 'Canada', 'Germany', 'France', 'Italy', 'Netherlands',
       'United Kingdom', 'United States'], dtype=object)

- Eight(8) different countries in the data: Australia, Canada, Germany, France, Italey, Netherlands, United Kingdom and United States.

In [196]:
customers_df['Country'] = pd.Categorical(
    values = customers_df['Country'],
    categories = ['Australia', 'Canada', 'Germany', 'France', 'Italy', 'Netherlands',
       'United Kingdom', 'United States'],
    ordered = True
)

In [197]:
# Categorizing Continent Column.
customers_df['Continent'].unique()

array(['Australia', 'North America', 'Europe'], dtype=object)

- Three(3) different continents in the data: Australia, North America and Europe.

In [198]:
customers_df['Continent'] = pd.Categorical(
    values = customers_df['Continent'],
    categories = ['Australia', 'North America', 'Europe'],
    ordered = True
)

In [199]:
# Checking progress on memory usage.
customers_df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15266 entries, 301 to 2099937
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   Gender      15266 non-null  category
 1   Name        15266 non-null  object  
 2   City        15266 non-null  object  
 3   State Code  15256 non-null  object  
 4   State       15266 non-null  object  
 5   Zip Code    15266 non-null  object  
 6   Country     15266 non-null  category
 7   Continent   15266 non-null  category
 8   Birthday    15266 non-null  object  
dtypes: category(3), object(6)
memory usage: 5.9 MB


- 5.9 MB of memory usage. 

In [228]:
# Checking to see if Zip Code column needs to be changed to int data type.
# customers_df.astype({'Zip Code': 'int64'}).dtypes

- Canada, Netherlands and United Kingdom utilize alpha-numeric zipcodes, this column must remain in the object data type.

In [201]:
# Changing Birthday Column to date time data type.
customers_df.astype({'Birthday': 'datetime64'}).dtypes

Gender              category
Name                  object
City                  object
State Code            object
State                 object
Zip Code              object
Country             category
Continent           category
Birthday      datetime64[ns]
dtype: object

In [202]:
# Final view of dataframe.
customers_df.head()

Gender               Name            City State Code  \
CustomerKey                                                         
301          Female      Lilly Harding  WANDEARAH EAST         SA   
325          Female       Madison Hull      MOUNT BUDD         WA   
554          Female      Claire Ferres       WINJALLOK        VIC   
786            Male  Jai Poltpalingada    MIDDLE RIVER         SA   
1042           Male    Aidan Pankhurst   TAWONGA SOUTH        VIC   

                         State Zip Code    Country  Continent    Birthday  
CustomerKey                                                                
301            South Australia     5523  Australia  Australia    7/3/1939  
325          Western Australia     6522  Australia  Australia   9/27/1979  
554                   Victoria     3380  Australia  Australia   5/26/1947  
786            South Australia     5223  Australia  Australia   9/17/1957  
1042                  Victoria     3698  Australia  Australia  11/19/1965

In [229]:
# Exporting the new data table.
# customers_df.to_csv('customers_df')

## Products Table

In [204]:
# Loading the Dataframe for 'Products'.
products_df = pd.read_csv('Products.csv', index_col='ProductKey', encoding = 'unicode_escape')
products_df.tail(5)

Product Name    Brand   Color  \
ProductKey                                                                   
2513          Contoso Bluetooth Active Headphones L15 Red  Contoso     Red   
2514        Contoso Bluetooth Active Headphones L15 White  Contoso   White   
2515                   Contoso In-Line Coupler E180 White  Contoso   White   
2516                   Contoso In-Line Coupler E180 Black  Contoso   Black   
2517                  Contoso In-Line Coupler E180 Silver  Contoso  Silver   

           Unit Cost USD Unit Price USD  SubcategoryKey  \
ProductKey                                                
2513             $43.07        $129.99              505   
2514             $43.07        $129.99              505   
2515              $1.71          $3.35              505   
2516              $1.71          $3.35              505   
2517              $1.71          $3.35              505   

                        Subcategory  CategoryKey     Category  
ProductKey                                                     
2513        Cell phones Accessories            5  Cell phones  
2514        Cell phones Accessories            5  Cell phones  
2515        Cell phones Accessories            5  Cell phones  
2516        Cell phones Accessories            5  Cell phones  
2517        Cell phones Accessories            5  Cell phones

In [205]:
# Reviewing the Dataframe.
products_df.info(memory_usage ='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2517 entries, 1 to 2517
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Product Name    2517 non-null   object
 1   Brand           2517 non-null   object
 2   Color           2517 non-null   object
 3   Unit Cost USD   2517 non-null   object
 4   Unit Price USD  2517 non-null   object
 5   SubcategoryKey  2517 non-null   int64 
 6   Subcategory     2517 non-null   object
 7   CategoryKey     2517 non-null   int64 
 8   Category        2517 non-null   object
dtypes: int64(2), object(7)
memory usage: 1.3 MB


- Dataframe Keys: Product Name, Brand, Color, Unit Cost USD, Unit Price USD, SubcategoryKey, Subcategory, CategoryKey and Category.
- 2517 entries.
- No null values.
- Only running on 196 KB, relatively small.
- Keys 3 and 4 are prices, candidates for float data type for quantitative analysis.
- Keys 5, 6, 7 and 8 candidates for Categorical data types. 

In [206]:
# Unique value totals.
products_df.nunique()

Product Name      2517
Brand               11
Color               16
Unit Cost USD      480
Unit Price USD     426
SubcategoryKey      32
Subcategory         32
CategoryKey          8
Category             8
dtype: int64

- Turns out Category/Subcategory keys are best candidates for Categorical data type.

In [207]:
# Checking for duplicated entries.
products_df.duplicated().sum()

0

In [208]:
# Checking memory usage.
products_df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2517 entries, 1 to 2517
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Product Name    2517 non-null   object
 1   Brand           2517 non-null   object
 2   Color           2517 non-null   object
 3   Unit Cost USD   2517 non-null   object
 4   Unit Price USD  2517 non-null   object
 5   SubcategoryKey  2517 non-null   int64 
 6   Subcategory     2517 non-null   object
 7   CategoryKey     2517 non-null   int64 
 8   Category        2517 non-null   object
dtypes: int64(2), object(7)
memory usage: 1.3 MB


In [209]:
# Chanwging CategoryKey and Category into Categorical data types.
products_df['Category'].unique()

array(['Audio', 'TV and Video', 'Computers', 'Cameras and camcorders',
       'Cell phones', 'Music, Movies and Audio Books', 'Games and Toys',
       'Home Appliances'], dtype=object)

In [210]:
products_df['Category'] = pd.Categorical(
    values = products_df['Category'],
    categories = ['Audio', 'TV and Video', 'Computers', 'Cameras and camcorders',
       'Cell phones', 'Music, Movies and Audio Books', 'Games and Toys',
       'Home Appliances'],
    ordered = True
)

In [211]:
products_df['CategoryKey'].unique()

array([1, 2, 3, 4, 5, 6, 7, 8], dtype=int64)

In [212]:
products_df['CategoryKey'] = pd.Categorical(
    values = products_df['CategoryKey'],
    categories = [1, 2, 3, 4, 5, 6, 7, 8],
    ordered = True
)

In [213]:
products_df.dtypes

Product Name        object
Brand               object
Color               object
Unit Cost USD       object
Unit Price USD      object
SubcategoryKey       int64
Subcategory         object
CategoryKey       category
Category          category
dtype: object

In [214]:
# Converting Unit Cost USD and Unit Price USD to floats for better quantitative calculations.
# First step is removing the non-numeric '$' and ',' from the the numbers.
products_df['Unit Cost USD'] = products_df['Unit Cost USD'].str.replace('$', '')
products_df['Unit Cost USD'] = products_df['Unit Cost USD'].str.replace(',', '')
products_df['Unit Price USD'] = products_df['Unit Price USD'].str.replace('$', '')
products_df['Unit Price USD'] = products_df['Unit Price USD'].str.replace(',', '')
products_df.head()

C:\Users\aaron\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\aaron\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  """


Product Name    Brand   Color  \
ProductKey                                                         
1           Contoso 512MB MP3 Player E51 Silver  Contoso  Silver   
2             Contoso 512MB MP3 Player E51 Blue  Contoso    Blue   
3              Contoso 1G MP3 Player E100 White  Contoso   White   
4             Contoso 2G MP3 Player E200 Silver  Contoso  Silver   
5                Contoso 2G MP3 Player E200 Red  Contoso     Red   

           Unit Cost USD Unit Price USD  SubcategoryKey Subcategory  \
ProductKey                                                            
1                  6.62          12.99              101     MP4&MP3   
2                  6.62          12.99              101     MP4&MP3   
3                  7.40          14.52              101     MP4&MP3   
4                 11.00          21.57              101     MP4&MP3   
5                 11.00          21.57              101     MP4&MP3   

           CategoryKey Category  
ProductKey                       
1                    1    Audio  
2                    1    Audio  
3                    1    Audio  
4                    1    Audio  
5                    1    Audio

In [215]:
# Now that we're dealing with just numbers, changing the data type to float.
products_df['Unit Cost USD'] = pd.to_numeric(products_df['Unit Cost USD'])
products_df['Unit Price USD'] = pd.to_numeric(products_df['Unit Price USD'])
products_df.dtypes

Product Name        object
Brand               object
Color               object
Unit Cost USD      float64
Unit Price USD     float64
SubcategoryKey       int64
Subcategory         object
CategoryKey       category
Category          category
dtype: object

 - Unit Cost USD and Unit Prices USD converted to float64 data type.

In [216]:
# Exporting the dataframe.

## Sales Table

In [217]:
# Loading Sales.csv into Dataframe.
sales_df = pd.read_csv('Sales.csv', index_col = 'ProductKey')
sales_df.tail()

Order Number  Line Item Order Date Delivery Date  CustomerKey  \
ProductKey                                                                  
632              2243030          1  2/20/2021           NaN      1216913   
98               2243031          1  2/20/2021     2/24/2021       511229   
1613             2243032          1  2/20/2021     2/23/2021       331277   
1717             2243032          2  2/20/2021     2/23/2021       331277   
464              2243032          3  2/20/2021     2/23/2021       331277   

            StoreKey  Quantity Currency Code  
ProductKey                                    
632               43         3           USD  
98                 0         4           EUR  
1613               0         2           CAD  
1717               0         2           CAD  
464                0         7           CAD

In [218]:
# Reviewing Dataframe.
sales_df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 62884 entries, 1304 to 464
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Order Number   62884 non-null  int64 
 1   Line Item      62884 non-null  int64 
 2   Order Date     62884 non-null  object
 3   Delivery Date  13165 non-null  object
 4   CustomerKey    62884 non-null  int64 
 5   StoreKey       62884 non-null  int64 
 6   Quantity       62884 non-null  int64 
 7   Currency Code  62884 non-null  object
dtypes: int64(5), object(3)
memory usage: 12.8 MB


- 8 columns with 62884 entries.
- Columns: Order Number, Line Number, Order Date, Delivery Date, CustomerKey, StoreKey, Quantity and Currency Code.

In [219]:
# Checking for duplicated rows.
sales_df.duplicated().sum()

0

In [220]:
# Checking for unique value counts.
sales_df.nunique()

Order Number     26326
Line Item            7
Order Date        1641
Delivery Date     1492
CustomerKey      11887
StoreKey            58
Quantity            10
Currency Code        5
dtype: int64

- Currency Code candidate for Categorical data type.

In [221]:
# Converting Currency Code into Categorical data type.
sales_df['Currency Code'].unique()

array(['CAD', 'USD', 'GBP', 'EUR', 'AUD'], dtype=object)

In [222]:
sales_df['Currency Code'] = pd.Categorical(
    values = sales_df['Currency Code'],
    categories = ['CAD', 'USD', 'GBP', 'EUR', 'AUD'],
    ordered = True
)

In [223]:
sales_df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 62884 entries, 1304 to 464
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   Order Number   62884 non-null  int64   
 1   Line Item      62884 non-null  int64   
 2   Order Date     62884 non-null  object  
 3   Delivery Date  13165 non-null  object  
 4   CustomerKey    62884 non-null  int64   
 5   StoreKey       62884 non-null  int64   
 6   Quantity       62884 non-null  int64   
 7   Currency Code  62884 non-null  category
dtypes: category(1), int64(5), object(2)
memory usage: 9.2 MB


- Improved memory usage by 3 MB.

In [224]:
sales_df['Delivery Date'] = pd.to_datetime(sales_df['Delivery Date'])
sales_df.dtypes

Order Number              int64
Line Item                 int64
Order Date               object
Delivery Date    datetime64[ns]
CustomerKey               int64
StoreKey                  int64
Quantity                  int64
Currency Code          category
dtype: object

In [225]:
sales_df['Order Date'] = pd.to_datetime(sales_df['Order Date'], errors='ignore')

In [226]:
sales_df.dtypes

Order Number              int64
Line Item                 int64
Order Date       datetime64[ns]
Delivery Date    datetime64[ns]
CustomerKey               int64
StoreKey                  int64
Quantity                  int64
Currency Code          category
dtype: object

In [227]:
# Exporting the Dataframe.